### DISCUSSION TITLE
Accessing the $validator property inside the $callback function of ClosureValidationRule

### DISCUSSION THREAD

**Discussion by shinsenter** [laravel/framework#49006] [Ideas] [answered]
## My problem

Hello everyone.

I'm creating a custom validation rule using Closure, and inside that rule I need to access other fields in the data array I'm validating.

A lot of articles suggest accessing fields from the request, but that won't work for me since I'm validating an array loaded from a CSV file inside a loop.

## Idea

I looked at the ClosureValidationRule source code, and it would be really helpful if the $validator property could be passed as a 4th argument to the $callback function. That way I could access the data array inside my custom rule.

https://github.com/laravel/framework/blob/10.x/src/Illuminate/Validation/ClosureValidationRule.php

```diff
    public function passes($attribute, $value)
    {
        $this->failed = false;

        $this->callback->__invoke($attribute, $value, function ($attribute, $message = null) {
            $this->failed = true;

            return $this->pendingPotentiallyTranslatedString($attribute, $message);
-        });
+        }, $this->validator);

        return ! $this->failed;
    }
```

## Usage

https://laravel.com/docs/10.x/validation#using-closures

```diff
use Closure;
use Illuminate\Support\Facades\Validator;
use Illuminate\Support\LazyCollection;
+ use Illuminate\Validation\Validator as ValidationValidator;

$rules = [
    'price' => [
-       function (string $attribute, mixed $value, Closure $fail) {
+       function (string $attribute, mixed $value, Closure $fail, ValidationValidator $validator) {
-           if ($value === 'foo') {
+           if (!price_logic($value, $validator->attributes()['payment_country'] ?? null)) {
                $fail("The {$attribute} is invalid.");
            }
        },
    ],
];

/** @var LazyCollection $csv_rows */
$csv_rows = next_row_from_csv();

foreach ($csv_rows as $row) {
    if (Validator::make($row, $rules)->failed()) {
        break;
    }
}
```

That's my idea. I'm a bit stuck and could use some help with this.
Let me know if you have any suggestions on how I can get access to the data array from within my custom rule callback.

Thanks!

**Comment by shinsenter** [comment]
> ```
> $validator->attributes()['payment_country'] ?? null
> ```

⇧
Being able to access values via a getValue() method on the Validator would be useful.

```diff
- $validator->attributes()['payment_country'] ?? null
+ $validator->getValue('payment_country')
```

This pull request may be relevant: #49007

**Comment by rojtjo** [comment]
You could use the [`DataAwareRule`](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Contracts/Validation/DataAwareRule.php) interface for this.

```php
use Illuminate\Contracts\Validation\DataAwareRule;
use Illuminate\Contracts\Validation\ValidationRule;
 
class ValidPrice implements DataAwareRule, ValidationRule
{
    protected $data = [];

    public function setData(array $data): static
    {
        $this->data = $data;
 
        return $this;
    }

    public function validate(string $attribute, mixed $value, Closure $fail): void
    {
        if (!price_logic($value, $this->data['payment_country'] ?? null)) {
            $fail("The {$attribute} is invalid.");
        }
    }
}
```

More info here under the `Accessing Additional Data` section
https://laravel.com/docs/10.x/validation#using-rule-objects

**Reply by shinsenter** [reply]
@rojtjo 

Thank you for suggesting to create a new ValidationRule.

The purpose of this discussion is that I just want the validation using Closure to be more flexible than it currently is. Creating many new ValidationRules for small validation snippets makes my Rules folder become messy and harder to manage.

Thanks again.

**Reply by rojtjo** [reply]
Ah I see your point!

You could however just create a generic implementation that wraps your closures like this:

```php
$rules = [
    'price' => new DataAwareValidationRule(function (string $attribute, mixed $value, Closure $fail, array $data) {
        if (!price_logic($value, $data['payment_country'] ?? null)) {
            $fail("The {$attribute} is invalid.");
        }
    }),
];
```

```php
use Closure;
use Illuminate\Contracts\Validation\DataAwareRule;
use Illuminate\Contracts\Validation\ValidationRule;

final class DataAwareValidationRule implements DataAwareRule, ValidationRule
{
    private Closure $validate;

    private array $data = [];

    public function __construct(Closure $validate)
    {
        $this->validate = $validate;
    }

    public function setData(array $data): self
    {
        $this->data = $data;

        return $this;
    }

    public function validate(string $attribute, mixed $value, Closure $fail): void
    {
        ($this->validate)($attribute, $value, $fail, $this->data);
    }
}
```

You could of course also create a similar class that implements [`ValidatorAwareRule`](https://github.com/laravel/framework/blob/10.x/src/Illuminate/Contracts/Validation/ValidatorAwareRule.php) if you want access to the full `Validator` instance. This might be helpful if your pull request (#49007) is accepted.

**Reply by shinsenter** [reply]
Hi @rojtjo 

Thank you once again for the solution of wrapping the Closure inside a ValidationRule. I also just thought of it a few hours ago and it worked as I expected.

However, I still think it would be better if we could pass the property $validator as a parameter to the Closure, so that we could take advantage of the existing $validator property in the class ClosureValidationRule.

**Comment by shinsenter** [comment, marked as answer]
Merged [#49015](https://github.com/laravel/framework/pull/49015) into 10.x.
